src/charm.py

```diff
@@ -157,10 +163,14 @@ def _update_config(self, event=None):
         try:
             if yaml.safe_load(config) != yaml.safe_load(old_config):
                 self._container.push(CONFIG_PATH, config)
+                self._stored.agent_config = config
                 # FIXME: #19
                 # self._reload_config()
                 self._container.restart(self._name)
                 self.unit.status = ActiveStatus()
+        except APIError as e:
+            self.unit.status = WaitingStatus(str(e))
+            event.defer()
```

Comment:
What does APIError mean here? I have some doubts that deferring is the correct action.

Reply:
When you kill a pod, Juju will create a new one to replace it.

When creating the new unit the `_on_loki_push_api_endpoint_joined` event is fired before `pebble-ready`event, BUT when Pebble is ready (that why `self._container.can_connect()` in line 149 is `True`).

`APIError` is raised because `agent` service doesn't exist yet since we add that layer in `on_pebble_ready`:

```
  File "./src/charm.py", line 88, in _on_loki_push_api_endpoint_joined
    self._update_config(event)
  File "./src/charm.py", line 169, in _update_config
    self._container.restart(self._name)
  File "/var/lib/juju/agents/unit-grafana-agent-k8s-0/charm/venv/ops/model.py", line 1159, in restart
    self._pebble.start_services(service_names)
  File "/var/lib/juju/agents/unit-grafana-agent-k8s-0/charm/venv/ops/pebble.py", line 1251, in start_services
    return self._services_action('start', services, timeout, delay)
  File "/var/lib/juju/agents/unit-grafana-agent-k8s-0/charm/venv/ops/pebble.py", line 1304, in _services_action
    resp = self._request('POST', '/v1/services', body=body)
  File "/var/lib/juju/agents/unit-grafana-agent-k8s-0/charm/venv/ops/pebble.py", line 1115, in _request
    response = self._request_raw(method, path, query, headers, data)
  File "/var/lib/juju/agents/unit-grafana-agent-k8s-0/charm/venv/ops/pebble.py", line 1155, in _request_raw
    raise APIError(body, code, status, message)
ops.pebble.APIError: cannot start services: service "agent" does not exist
```

That's why I'm catching that exception and deferring the event.

Reply:
> When creating the new unit the `_on_loki_push_api_endpoint_joined` event is fired before `pebble-ready`event

I would put that in a comment.

Reply:
done